acceptancetests/assess_upgrade.py

```diff
@@ -0,0 +1,199 @@
+    client.get_controller_client().juju('upgrade-juju', ('-m', 'controller'))
+    assert_model_is_version(client.get_controller_client(), devel_version)
+    wait_until_model_upgrades(client)
+
+    client.juju('upgrade-juju', ())
+    assert_model_is_version(client, devel_version)
+    wait_until_model_upgrades(client)
+
+
+def wait_until_model_upgrades(client, timeout=300):
+    # Poll using a command that will fail until the upgrade is complete.
+    def model_upgrade_status_check(client):
+        try:
+            log.info('Attempting API connection, failure is not fatal.')
+            client.list_models()
+            client.juju('list-users', (), include_e=False)
+            return True
+        except CalledProcessError:
+            # Upgrade will still be in progress and thus refuse the api call.
+            return False
+    try:
+        wait_for_model_check(client, model_upgrade_status_check, timeout)
+    except ModelCheckFailed:
+        raise JujuAssertionError(
+            'Upgrade for model {} failed to complete within the alloted '
+            'timeout ({} seconds)'.format(
+                client.model_name, timeout))
+
+
+def assert_stable_model_is_correct(stable_client):
+    assert_model_is_version(
+        stable_client.get_controller_client(),
+        get_stripped_version_number(stable_client.version))
+    assert_model_is_version(
+        stable_client,
+        get_stripped_version_number(stable_client.version))
+
+
+def assert_model_is_version(client, expected_version, timeout=600):
+    # Check agent & machine version
+    client.wait_for_version(expected_version, timeout)
+    # Also check model version
+    client.wait_for(WaitModelVersion(expected_version, timeout))
+
+
+def setup_agent_metadata(stream_server, agent, client, tmp_dir, stream):
+    version_parts = get_version_parts(client.version)
+
+    if agent is None:
+        agent_details = agent_tgz_from_juju_binary(client.full_path, tmp_dir)
+    else:
+        agent_details = agent
+
+    stream_server.add_product(
+        stream,
+        version_parts.version,
+        version_parts.arch,
+        client.env.get_option('default-series'),
+        agent_details)
+    # Trusty needed for wikimedia charm.
+    stream_server.add_product(
+        stream,
+        version_parts.version,
+        version_parts.arch,
+        'trusty',
+        agent_details)
+
+
+def get_version_parts(version_string):
+    parts = version_string.split('-')
+    return VersionParts(parts[0], parts[1], parts[2])
+
+
+def parse_args(argv):
+    """Parse all arguments."""
+    parser = argparse.ArgumentParser(
+        description='Test juju upgrades for controllers and models')
+    add_basic_testing_arguments(parser, existing=False)
+    parser.add_argument(
+        '--stable-juju-bin',
+        help='Path to juju binary to be used as the stable version of juju.')
+    parser.add_argument(
+        '--stable-agent',
```

Comment:
--stable-juju-agent to mirror --stable-juju-bin?

Reply:
Good point, changed.